# Promise

一般而言，在写Java或Android的时候，若要请求一些需要长时间等待的资源，我们是怎么办的呢？比如文件读取，我们都会启动一条新的线程，新线程发出文件读取的请求后停下来等待；直到系统把硬盘的数据读取到内存，新线程再把数据送给主线程；在这整个过程中，主线程是不受到什么影响的，该干嘛干嘛，也可以随时响应用户的操作。

可是JavaScript不行啊，它是一门单线程语言啊，它不能像其它语言一样，遇到耗时任务就启动新线程。它所有的任务都必须在一条线程内完成，如果按照同步的方式处理用户肯定受不了，所以我们给它配上了异步的能力：为耗时任务注册事件处理函数，发送请求后把事件处理函数丢到一个事件队列中等待，自己继续干后面的事；当耗时任务结束时，它再把任务返回的数据作为参数调用事件处理函数。

So，有过一些前端开发的经验可能都会写出下面这样的代码：

````
$.ajax({
    url: 'xxx',
    ...
    success: function(res) {
        ...
        $.ajax({
            url: 'yyy',
            ...
            success: function(res2) {
                ...
                $.ajax({
                    url: 'www',
                    ...
                    success: function(res3) {
                        $.ajax({
                            url: 'zzz',
                            ...
                            success: function(res4) {
                                ...
                            }
                        });
                    }
                });
            }
        });
    }
});
````

像这样一个箭头走到黑的代码，被称为“回调地狱”，不仅难看、难调试，还容易造成BUG或内存泄漏等问题。很早之前就有人想改变这种写法了，最早的promise也是在那时出世的，当然那时的它还只是一个库而不是现在的语言标准的Promise。如果把上面的代码改成Promise会是什么样子的呢？

````
fetch('xxx')
.then((res) => {
    ...
    return fetch('yyy');
})
.then((res2) => {
    ...
    return fetch('www');
})
.then((res3) => {
    ...
    return fetch('zzz');
})
.then((res4) => {
    ...
});
````

它直接按事件的顺序写成一条then链，这样看起来舒服多了吧！所以说到底，Promise是异步编程的一种解决方案，相比传统的回调函数与事件更加合理和强大。

## 基本概念与用法

所谓Promise，简单说就是一个容器，里面保存着某个未来才会结束的事件（通常是一个异步操作）的结果。从语法上说，Promise 是一个对象，从它可以获取异步操作的消息，这个对象有以下两个特点：

1. 对象代表一种异步操作，有三种状态：padding、fulfilled、rejected。只有异步操作的结果可以决定当前的状态，任何其它操作都无法再改变。
2. 状态只能从padding->fulfilled或padding->rejected，一旦状态改变，就不会再变，任何时候都可以得到这个结果。

当然从这特点也可以看出些许Promise的缺点：一旦开始就会立即执行，中途无法取消；若不设置回调函数，内部抛出的错误不会反应到外部；处于padding状态时，无法得知进展。个人感觉这些缺点有时候也不完全算是缺点吧，后面再讨论！现在来看看怎么使用Promise。

ES6规定，Promise对象是一个构造函数，用来生成Promise实例，这个构造函数接受一个函数作为参数，而该函数的两个参数分别是resolve(fulfilled)和reject。这两个参数也是函数来的，不过不用自己定义，由引擎提供，分别让你在异步操作成功完成或失败时调用。如下：

````
const p = new Promise((resolve, reject) => {
    ...
    if (success) {
        resolve(dataFromAsyncRequest);
    } else {
        reject(errorMessage);
    }
});
````

实例生成后，可用then和catch分别指定resolved状态和rejected状态的回调函数，回调函数分别接受一个参数，这个参数的值分别就是调用resolve函数或reject函数传进去的值（其实then可以接受两个回调函数，第一个作用不变，第二个与catch的回调函数相同，但是这种写法太难看了）。下面代码接上文：

````
p.then((data) => {  // data === dataFromAsyncRequest
    ...
}).catch((e) => {  // e === errorMessage
    ...
});
````

其实在上面例子，一旦Promise被实例化，传入Promise的函数会立即执行，而then或catch不会立即执行，它们会等函数内操作（异步或同步都一样）完成并且resolve或reject被调用之后才执行。所以一般来说，调用resolve或reject之后，Promise的使命就结束了，但是若调用它们的语句后面还有语句，那些语句会继续执行下去的，而不是像return一样跳出函数。这是因为立即resolved的Promise是在本轮事件循环的末尾执行，总是晚于本轮循环的同步任务。所以啊，我们在执行resolve或reject时，最好还是在它们前面加个return吧！

## 链式调用

then和catch方法都是定义在原型对象Promise.prototype上的，其实啊，它们都会返回一个新的Promise实例，所以它们才可以像上面一样使用链式调用的方法（有点类似jQuery的DOM操作）。这时，前一个回调函数，有可能返回的还是一个Promise对象（即有异步操作），这时后一个回调函数，就会等待该Promise对象的状态发生变化，才会被调用。所以，我们才能对一系列的异步操作按某种次序调用。

下面可以看到，在then的回调函数中返回了什么，下一个then的回调函数就会接收到什么；而若返回的是一个新的Promise，那要等这个新Promise执行完成之后，再把新Promise中传入resolve或reject函数的值送到下一个then或catch方法的回调函数中。我们就是用这种方法避免回调地狱的：

````
// 在Promise对象内，应该要调用res或rej之一，传给它们的参数将会传到后面的then或catch中
new Promise((res, rej) => {
    res(1);
})
.then((data) => {  // data === 1
    return 2;  // 在then中返回的参数又会被传到后面的then
})
.then((data) => {  // data === 2
    // 如果在then中返回一个新的Promise，那下面的then或catch响应的是新Promise中的res或rej(throw)
    return new Promise((res, rej) => {
        res(3);
    });
}).then((data) => {  // data === 3
    console.log(`finished`);
    throw 4;
}).catch((err) => {  // err === 4
	...
});
````

如果在一条链中，没有出错，所有then方法都会被依序执行下去，而catch会被跳过；若Promise对象内先调用了reject或者在任何地方抛出了错误，那最近的catch方法才会执行。而由于Promise一旦开启就不会中途取消，因此Promise的then/catch链不会中断，必须从头到尾执行下去。如果我们要跳出多个连续then，只能使用reject()或throw抛出异常，然后让最近一个catch方法捕获它。

````
new Promise((res, rej) => {
    res(1);
}).then((data) => {
    if (failed) {  // 若在这里不想执行下面的then了，抛出一个异常即可
        throw 9;
    }
    return 2;
}).then((data) => {
    ...
}).then((data) => {
    ...
}).catch((err) => {
    if (err === 9) {  // 其实可以自己定义一种Error，以便catch判断是哪个then抛出的
        ...
    }
    ...  // 后面还可以继续工作，或者继续接上then/catch
});
````

## 了解一下

- Promise.prototype.finally()
- Promise.all()
- Promise.race()
- Promise.resolve()
- Promise.reject()
- Promise.try()


## 参考链接

- [Promise对象](http://es6.ruanyifeng.com/#docs/promise#%E5%BA%94%E7%94%A8)